Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,188 @@
+            return 4
+        case .address(.v6):
+            return 16
+        case .address(.unixDomainSocket):
+            fatalError("Unsupported")
+        case .domain(let domain, port: _):
+            return domain.count + 1
+        }
+    }
+}
+
+extension ByteBuffer {
+    
+    mutating func readAddressType() throws -> AddressType? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard let type = buffer.readInteger(as: UInt8.self) else {
+                throw MissingBytes()
+            }
+            
+            switch type {
+            case 0x01:
+                return try buffer.readIPv4Address()
+            case 0x03:
+                return try buffer.readDomain()
+            case 0x04:
+                return try buffer.readIPv6Address()
+            default:
+                throw SOCKSError.InvalidAddressType(actual: type)
+            }
+        }
+    }
+    
+    mutating func readIPv4Address() throws -> AddressType? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard
+                let bytes = buffer.readSlice(length: 4),
+                let port = try buffer.readPort()
+            else {
+                throw MissingBytes()
+            }
+            return .address(try .init(packedIPAddress: bytes, port: port))
+        }
+    }
+    
+    mutating func readIPv6Address() throws -> AddressType? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard
+                let bytes = buffer.readSlice(length: 16),
+                let port = try buffer.readPort()
+            else {
+                throw MissingBytes()
+            }
+            return .address(try .init(packedIPAddress: bytes, port: port))
+        }
+    }
+    
+    mutating func readDomain() throws -> AddressType? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard
+                let length = buffer.readInteger(as: UInt8.self),
+                let host = buffer.readString(length: Int(length)),
+                let port = try buffer.readPort()
+            else {
+                throw MissingBytes()
+            }
+            return .domain(host, port: UInt16(port))
+        }
+    }
+    
+    mutating func readPort() throws -> Int? {
+        return self.readInteger(as: UInt16.self).flatMap { Int($0) }
+    }
+    
+    @discardableResult mutating func writeAddressType(_ type: AddressType) -> Int {
+        switch type {
+        case .address(.v4(let address)):
+            return self.writeInteger(UInt8(1))
+                + self.writeInteger(address.address.sin_addr.s_addr, endianness: .little)
+                + self.writeInteger(address.address.sin_port, endianness: .little)
+        case .address(.v6(let address)):
+            return self.writeInteger(UInt8(4))
+                + self.writeIPv6Address(address.address)
+                + self.writeInteger(address.address.sin6_port, endianness: .little)
```

Comment:
This endianness is wrong, it should be `.big`. The same applies to the rest of the explicit endianness code in this function.